### PR TITLE
list all images that have are derived from the given image

### PR DESCRIPTION
**Pull request by aokirisaki**
Signed-off-by: Lisca Ana-Roberta <ana.kagome@yahoo.com>

**What type of PR is this?**

**Which issue does this PR fix**:
#660 

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by peusebiu — COMMENTED** [review]
remove things under hack/tools
remove regctl, zot-0.1.0.tgz and pushpull_output.txt?

File: hack/tools/lib/bats-core/common.bash [hunk 1]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env bash
```
Comment (andaaron) [inline hack/tools/lib/bats-core/common.bash:1]: Remove the hack folder from the commit. See Sebi's comment.

File: pkg/extensions/search/schema.graphql [hunk 2]
```diff
@@ -19,54 +19,11 @@ type PackageInfo {
      FixedVersion: String 
 }
 
-type ImgResultForCVE {
```
Comment (andaaron) [inline pkg/extensions/search/schema.graphql:22]: Let's do this cleanup as part of https://github.com/project-zot/zot/pull/457.

In this PR, implement just https://github.com/project-zot/zot/pull/696/files#diff-84af1556304ab2f0edbb3541ee90f5e8eb88e4dde65c0a8069e517dea758c210R145, that's the new name of the query you implemented.

Clenan schema.graphql, add that query, and regenerate the code using gqlgen

File: pkg/extensions/search/schema.resolvers.go [hunk 3]
```diff
@@ -419,7 +434,118 @@ func (r *queryResolver) GlobalSearch(ctx context.Context, query string) (*gql_ge
 	}, nil
 }
 
+// DependencyListForImage is the resolver for the DependencyListForImage field.
+func (r *queryResolver) DependencyListForImage(ctx context.Context, image string) ([]*gql_generated.ImageSummary, error) {
+	defaultStore := r.storeController.DefaultStore
+	imageList := make([]*gql_generated.ImageSummary, 0)
+
+	repoList, err := defaultStore.GetRepositories()
```
Comment (andaaron) [inline pkg/extensions/search/schema.resolvers.go:442]: We need to look into all stores. I think we have this problem in other queries as well.
We should have a new method in layoutUtils which returns all repos in all stores.

**Review by andaaron — COMMENTED** [review]
Remove the files which are not supposed to be part of the review (zot-0.1.0.tgz, regctl, pushpull_output.txt)

Also I suggest a more detailed commit message.

Will circle back to this PR after the existing comments are addressed.

File: pkg/extensions/search/schema.resolvers.go [hunk 4]
```diff
@@ -463,6 +463,74 @@ func (r *queryResolver) GlobalSearch(ctx context.Context, query string) (*gql_ge
...
+
+		// verify every image
+		for _, manifest := range manifests {
+			layers := manifest.Layers
+
+			sameLayer := 0
+
+			for _, l := range imageLayers {
+				for _, k := range layers {
+					if *k.Digest == string(l.Digest)[7:] {
```
Comment (andaaron) [inline pkg/extensions/search/schema.resolvers.go:513]: Let's use a clean way to do tics check.
layers are of type https://github.com/opencontainers/image-spec/blob/bc9c4bd9b2a4ec6f49f956e193f9c699d854507d/specs-go/v1/descriptor.go#L27
Which is of type Digest, and has various useful methods: https://github.com/opencontainers/go-digest/blob/master/digest.go#L37

Why would we do this comparison relative to the arbitrary characters starting from 7?

File: pkg/extensions/search/schema.resolvers.go [hunk 5]
```diff
@@ -463,6 +463,74 @@ func (r *queryResolver) GlobalSearch(ctx context.Context, query string) (*gql_ge
 	}, nil
 }
 
+// DependencyListForImage is the resolver for the DependencyListForImage field.
+func (r *queryResolver) DerivedImageList(ctx context.Context, image string) ([]*gql_generated.ImageSummary, error) {
```
Comment (andaaron) [inline pkg/extensions/search/schema.resolvers.go:467]: We need to make sure all of the layers in the image sent as parameter are layers in the images returned by this method.
We don't what to know if the images have the exact same layers.
We want images with layers which are a superset of the set of layers sent in the image sent as parameter.

So if image `A` given as parameter has 3 layers `a b c`, and  image `B` has 4 layers `a b c d`, B meets the condition to be derived from A.
So if image `A` given as parameter has 3 layers `a b c`, and  image `C` has 2 layers `a b`, B meets the condition to be a base image for A (good for your other PR for BaseImageList), but doesn't meet the condition for this query.

File: pkg/extensions/search/schema.resolvers.go [hunk 6]
```diff
@@ -463,6 +463,74 @@ func (r *queryResolver) GlobalSearch(ctx context.Context, query string) (*gql_ge
...
+					if *k.Digest == string(l.Digest)[7:] {
+						sameLayer++
+
+						break
+					}
+				}
+			}
+
+			// if all layers are the same
+			if sameLayer == len(imageLayers) {
```
Comment (andaaron) [inline pkg/extensions/search/schema.resolvers.go:522]: I think this should be `>=` and we should test for returning images which include all of the common layers, but also extra layers besides them.

**Review by andaaron — APPROVED** [review]

**Review by andaaron — CHANGES_REQUESTED** [review]
See my previous review.

File: pkg/extensions/search/schema.resolvers.go (hunk 4 above)
Comment (aokirisaki) [reply pkg/extensions/search/schema.resolvers.go:513]: in order to compare a digest with "sha256:" in front with one which doesn't have it (this is the way it is provided by the function I'm calling to get the digest)

File: pkg/extensions/search/schema.resolvers.go (hunk 5 above)
Comment (aokirisaki) [reply pkg/extensions/search/schema.resolvers.go:467]: I added tests to verify all possible cases, equal layers, less layers than A and more layers than A.

**Review by andaaron — APPROVED** [review]

**Comment by peusebiu** [comment]
LGTM

**Review by andaaron — DISMISSED** [review]
Looks good.

**Comment by rchincha** [comment]
@aokirisaki can you pls rebase and push this again. Too many merge conflicts.

**Review by andaaron — APPROVED** [review]

**Pull request merged**